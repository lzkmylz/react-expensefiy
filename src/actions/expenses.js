import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD EXPENSE
export const addExpense = (expense) => (
    {
        type: "ADD_EXPENSE",
        expense: expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {
            description,
            note,
            amount,
            createdAt
        };
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => (
    {
        type: "REMOVE_EXPENSE",
        id
    }
);

//START REMOVE EXPENSE
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const updates = {};
        updates[`users/${uid}/expenses/${id}`] = null;
        return database.ref().update(updates).then(() => {
            dispatch(removeExpense({ id }));
        });
    };
}

//EDIT EXPENSE
export const editExpense = (id, updates) => (
    {
        type: "EDIT_EXPENSE",
        id,
        updates
    }
);

//START EDIT EXPENSE
export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

//SET EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expensesData = [];
            snapshot.forEach((childSnapshot) => {
                expensesData.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });
            dispatch(setExpenses(expensesData));
        });
    };
};