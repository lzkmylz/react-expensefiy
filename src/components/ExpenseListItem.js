import React from 'react';
import { Link } from 'react-router-dom'

export const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <p>description: {description}</p>
        </Link>
        <p>amount: {amount}</p>
        <p>createdAt: {createdAt}</p>
    </div>
);

export default ExpenseListItem;