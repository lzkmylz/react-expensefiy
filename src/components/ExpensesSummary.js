import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expense';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
    const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses';
    const formatedExpensesTotal = numeral(selectExpensesTotal(props.expenses)).format('$0,0.00');

    return (
        <div className="page-header">
            <div className="content_container">
                <h1 className="page-header__title">
                    Viewing <span>{props.expenses.length}</span> {expenseWord} totaling <span>{formatedExpensesTotal}</span>
                </h1>
                <div className="page-header_actions">
                    <Link className="button" to="/create" >Add Expense</Link>
                </div>
            </div>
        </div>
    )
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);