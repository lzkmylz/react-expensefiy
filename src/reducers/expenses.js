// EXPENSE REDUSER

const expenseReduserDefaultState = [];

const expenseReduser = (state = expenseReduserDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];     
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                //console.log(expense.id, action.id)
                if (expense.id === action.id) {
                    //console.log(expense)
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                };
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};

export default expenseReduser;