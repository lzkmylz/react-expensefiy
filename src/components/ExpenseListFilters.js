import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'
import uuid from 'uuid';

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        };
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        let sortBy = e.target.value;
        if(sortBy === 'date') {
            this.props.sortByDate();
        } else if(sortBy === 'amount') {
            this.props.sortByAmount();
        }
    };
    

    render() {
        return (
            <div className="content_container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                        className="text-input" 
                        type="text" 
                        value={this.props.filters.text} 
                        onChange={this.onTextChange} 
                        placeholder="Search Expenses"
                        />
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange} >
                        <option value="date" >Date</option>
                        <option value="amount" >Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        startDateId={uuid()}
                        endDate={this.props.filters.endDate}
                        endDateId={uuid()}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);