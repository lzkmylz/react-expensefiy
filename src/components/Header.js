import React from 'react';
import { NavLink } from 'react-router-dom';
import { startLogOut } from '../actions/auth';
import { connect } from 'react-redux';


export const Header = ({ startLogOut }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/dashboard" activeClassName="is-active" > Home </NavLink>
        <NavLink to="/create" activeClassName="is-active" > Create </NavLink>
        <NavLink to="/edit" activeClassName="is-active" > Edit </NavLink>
        <NavLink to="/help" activeClassName="is-active" > Help </NavLink>
        <button onClick={startLogOut} >LogOut</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogOut: () => {
        dispatch(startLogOut());
    }
});

export default connect(undefined, mapDispatchToProps)(Header);