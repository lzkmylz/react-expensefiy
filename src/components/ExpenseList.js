import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expense';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobil">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <h1>ExpenseList</h1>
        {
            props.expenses.length === 0 ? (<p>No Expense Found!</p>) : 
            (props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />
            }))
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);