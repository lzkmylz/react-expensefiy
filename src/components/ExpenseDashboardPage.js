import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
        <ExpenseSummary />
    </div>
);

export default ExpenseDashboardPage;