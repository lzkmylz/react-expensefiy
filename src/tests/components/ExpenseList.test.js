import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render expenselist with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseList with no message error', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});