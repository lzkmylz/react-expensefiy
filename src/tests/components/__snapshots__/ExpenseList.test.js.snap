// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with no message error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          ExpenseList
        </h1>,
        <p>
          No Expense Found!
        </p>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "ExpenseList",
        },
        "ref": null,
        "rendered": "ExpenseList",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No Expense Found!",
        },
        "ref": null,
        "rendered": "No Expense Found!",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            ExpenseList
          </h1>,
          <p>
            No Expense Found!
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "ExpenseList",
          },
          "ref": null,
          "rendered": "ExpenseList",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No Expense Found!",
          },
          "ref": null,
          "rendered": "No Expense Found!",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render expenselist with expenses 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExpenseList
    expenses={
      Array [
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": 1,
          "note": "",
        },
        Object {
          "amount": 19500,
          "createdAt": -259200000,
          "description": "Rent",
          "id": 2,
          "note": "",
        },
        Object {
          "amount": 200,
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": 3,
          "note": "",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          ExpenseList
        </h1>,
        Array [
          <ExpenseListItem
            amount={195}
            createdAt={0}
            description="Gum"
            id={1}
            note=""
          />,
          <ExpenseListItem
            amount={19500}
            createdAt={-259200000}
            description="Rent"
            id={2}
            note=""
          />,
          <ExpenseListItem
            amount={200}
            createdAt={345600000}
            description="Credit Card"
            id={3}
            note=""
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "ExpenseList",
        },
        "ref": null,
        "rendered": "ExpenseList",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": 1,
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "amount": 19500,
          "createdAt": -259200000,
          "description": "Rent",
          "id": 2,
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "function",
        "props": Object {
          "amount": 200,
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": 3,
          "note": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            ExpenseList
          </h1>,
          Array [
            <ExpenseListItem
              amount={195}
              createdAt={0}
              description="Gum"
              id={1}
              note=""
            />,
            <ExpenseListItem
              amount={19500}
              createdAt={-259200000}
              description="Rent"
              id={2}
              note=""
            />,
            <ExpenseListItem
              amount={200}
              createdAt={345600000}
              description="Credit Card"
              id={3}
              note=""
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "ExpenseList",
          },
          "ref": null,
          "rendered": "ExpenseList",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "amount": 195,
            "createdAt": 0,
            "description": "Gum",
            "id": 1,
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "amount": 19500,
            "createdAt": -259200000,
            "description": "Rent",
            "id": 2,
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "function",
          "props": Object {
            "amount": 200,
            "createdAt": 345600000,
            "description": "Credit Card",
            "id": 3,
            "note": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
